### ISSUE TITLE
issues with ContainedMathbox in 0.0.7, 0.0.8, 0.0.9-dev-3

### ISSUE THREAD

**Issue by sritchie** [ChristopherChudzicki/mathbox-react#25]
When I try to get this running in Clojurescript I see, with `ContainedMathbox`:

```
react_devtools_backend.js:3973 Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.
    at eval (out/cljs-runtime/module$node_modules$mathbox_react$build$cjs$index.js:3:135)
    at eval (out/cljs-runtime/module$node_modules$mathbox_react$build$cjs$index.js:3:135)
    at eval (out/cljs-runtime/module$node_modules$mathbox_react$build$cjs$index.js:3:135)
    at eval (out/cljs-runtime/module$node_modules$mathbox_react$build$cjs$index.js:7:376)
    at div
    at eval (out/cljs-runtime/module$node_modules$mathbox_react$build$cjs$index.js:9:12)
    at cmp (out/cljs-runtime/reagent.impl.component.js:499:43)
    at cmp (out/cljs-runtime/reagent.impl.component.js:499:43)
    at cmp (out/cljs-runtime/reagent.impl.component.js:499:43)
    at cmp (out/cljs-runtime/reagent.impl.component.js:499:43)
    at cmp (out/cljs-runtime/reagent.impl.component.js:499:43)
    at cmp (out/cljs-runtime/reagent.impl.component.js:499:43)
    at cmp (out/cljs-runtime/reagent.impl.component.js:499:43)
    at div
    at div
    at cmp (out/cljs-runtime/reagent.impl.component.js:499:43)
    at cmp (out/cljs-runtime/reagent.impl.component.js:499:43)
    at cmp (out/cljs-runtime/reagent.impl.component.js:499:43)
overrideMethod @ react_devtools_backend.js:3973
printWarning @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:2
error$jscomp$0 @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:1
scheduleUpdateOnFiber @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:469
dispatchAction @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:296
eval @ out/cljs-runtime/module$node_modules$mathbox_react$build$cjs$index.js:4
invokePassiveEffectCreate @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:534
callCallback @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:653
invokeGuardedCallbackImpl @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:655
invokeGuardedCallback @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:56
flushPassiveEffectsImpl @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:537
exports.unstable_runWithPriority @ out/cljs-runtime/module$node_modules$scheduler$cjs$scheduler_development.js:14
runWithPriority$1 @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:161
flushPassiveEffects @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:533
eval @ out/cljs-runtime/module$node_modules$react_dom$cjs$react_dom_development.js:529
flushWork @ out/cljs-runtime/module$node_modules$scheduler$cjs$scheduler_development.js:5
channel.port1.onmessage @ out/cljs-runtime/module$node_modules$scheduler$cjs$scheduler_development.js:10
```

**Comment by sritchie** [comment]
I can get more on this later...

**Comment by sritchie** [comment]
Here is the component:

```
[box/ContainedMathbox
 {:options {:plugins ["core" "controls" "cursor"]
            :controls {:klass box/OrbitControls}
            :camera {}}
  :containerStyle {:height "400px" :width "100%"}}
 [box/Cartesian
  {:range [[0 1] [0 1] [0 1]]
   :scale [1 1 1]}
  [box/Volume
   {:id id
    :width width-rez
    :bufferWidth width-rez
    :height height-rez
    :bufferHeight height-rez
    :depth depth-rez
    :bufferDepth depth-rez
    :items 1
    :channels 4
    :live false
    :expr (fn [emit x y z]
            (emit x y z opacity))}
   [box/Point
    {:points (str "#" id)
     :colors (str "#" id)
     :color 0xffffff
     :size size}]]]]
```

**Comment by ChristopherChudzicki** [comment]
OK, I was able to reproduce this in just js. Interestingly, only in react 17 not in react 18.

But even in react 18, too many renders are happening. **As a workaround:** You can avoid the extra renders by not inlining `mathboxOptions`...either define it outside the component or inside the component via `useMemoize(() => options, [])`. 

React is always going to re-render your component tree if you inline mathboxOptions that way, but `mathbox-react` does not need to destroy and create an entire mathbox instance whenever that happens. So I'll work on that, which I suspect will fix the issue.

**Comment by sritchie** [comment]
Not recreating a mathbox instance will be a huge improvement, I think. Thanks for looking into this!

**Comment by ChristopherChudzicki** [comment]
Oops, this got auto-closed when #27 got merged.

So I think the real issue here was not with `mathboxOptions` causing unnecessary re-renders, it was with Point being a child of Volume. Volume should not have children. Previously mathbox-react would allow you to write
```xml
<Cartesian>
  <Volume>
    <Point />
  </Volume>
</Cartesian>
```
Then
- mathbox-react thought that Point's parent was Volume
- mathbox thought Point's parent was Cartesian (Volumes cannot have children)

And there's a line in mathbox-react that checks that react + mathbox both have the same concept of parent, and re-renders if they do not. (See https://github.com/ChristopherChudzicki/mathbox-react/commit/0c8eff2165edf168f9c3dadc662ee93db8104a7c for why).

So since the two concepts of parent were never the same, mathbox-react kept re-rendering.

The following nodes can have children:
```js
const CAN_HAVE_CHILDREN = [
  "view",
  "cartesian",
  "cartesian4",
  "polar",
  "spherical",
  "stereographic",
  "stereographic4",
  "transform",
  "transform4",
  "vertex",
  "fragment",
  "layer",
  "mask",
  "group",
  "inherit",
  "root",
  "unit",
  "rtt",
  "clock",
  "now",
  "move",
  "present",
  "reveal",
  "slide",
] 
```

I'm going to do
> React is always going to re-render your component tree if you inline mathboxOptions that way, but mathbox-react does not need to destroy and create an entire mathbox instance whenever that happens. So I'll work on that, which I suspect will fix the issue.
then I'll release the fix on npm

**Comment by sritchie** [comment]
Ah!

I had thought that point was supposed to be a child of volume because then point knows to attach to every xyz in the volume. Does it attach instead to the closest (on the left) data layer at the same level? This is good stress on my mental model.

**Comment by ChristopherChudzicki** [comment]
> Does it attach instead to the closest (on the left) data layer at the same level?

@sritchie I believe that's the default if you do not specify `points`. 
- Each mathbox primitive (Point, Line, Area, Cartesian, etc) has a bunch of "traits". The traits specify groups of properties that can be specified on that primitive.
- The `Point` primitive (along with `Line` and `Surface` and probably some others) include the "geometry" trait, see [src/primitives/types/draw/point.js#L22](https://github.com/unconed/mathbox/blob/master/src/primitives/types/draw/point.js#L22)
- The `"geometry"` trait has includes an attribute `"points"` defined as `Types.select()` (in [src/primitives/types/traits.ts#L187](https://github.com/unconed/mathbox/blob/master/src/primitives/types/traits.ts#L187))
- `Types.elect()` has a default value of `<` defined in [src/primitives/types/types.js#L352](https://github.com/unconed/mathbox/blob/master/src/primitives/types/types.js#L352)

and I believe `"<"` means "previous sibling".

You can also specify "points" (and colors) as 
- css selectors (like you did, passing the id)
- a selection...
    ```
    const selection = mathbox.select('volume');
    cartesian.point({ points: selection })
    ```
- or a node
    ```
    const node = mathbox.select('volume')[0];
    cartesian.point({ points: selection })
    ```

Re
> I had thought that point was supposed to be a child of volume because then point knows to attach to every xyz in the volume. 
That's not the mathbox API, but I did consider doing something like that as the mathbox-react API because it's a little weird in react to be passing string ids around (and getting references to the selections via `ref` is a bit tedious). I decided against doing that because in general `points` and `colors` could have different data sources... so which should should be inherited from parent? So I decided just to stick with the way MathBox does it.

**Comment by ChristopherChudzicki** [comment]
Resolved by #27
